File: leetcode/Triangle/minimumTotal.java
public class Solution {
    public int minimumTotal(ArrayList<ArrayList<Integer>> triangle) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int len = triangle.size(), sols[] = new int[len], ans = Integer.MAX_VALUE;
        for (int i = 0; i < len; i++) {
            for (int j = i; j >= 0; j--) {
                sols[j] = triangle.get(i).get(j) + (i == 0 ? 0 : Math.min(
                    j == 0 ? Integer.MAX_VALUE : sols[j - 1],
                    j == i ? Integer.MAX_VALUE : sols[j]));
                if (i == len - 1 && sols[j] < ans) ans = sols[j];
            }
        }
        return ans;
    }
}

File: leetcode/Triangle/minimumTotal.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size(), i = n - 2;
        while (i >= 0) {
            for (int j = 0; j < triangle[i].size(); j++) {
                triangle[i][j] += min(triangle[i + 1][j], triangle[i + 1][j + 1]);
            }
            i--;
        }
        return n ? triangle[0][0] : 0;
    }
};
